cartography/intel/azure/cosmosdb.py

```diff
@@ -0,0 +1,1090 @@
+    d.keyvaulturi = da.key_vault_key_uri,
+    d.enablefreetier = da.enable_free_tier,
+    d.enableanalyticalstorage = da.enable_analytical_storage,
+    d.defaultconsistencylevel = da.consistency_policy.default_consistency_level,
+    d.maxstalenessprefix = da.consistency_policy.max_staleness_prefix,
+    d.maxintervalinseconds = da.consistency_policy.max_interval_in_seconds
+    WITH d
+    MATCH (owner:AzureSubscription{id: {AZURE_SUBSCRIPTION_ID}})
+    MERGE (owner)-[r:RESOURCE]->(d)
+    ON CREATE SET r.firstseen = timestamp()
+    SET r.lastupdated = {azure_update_tag}
+    """
+
+    neo4j_session.run(
+        ingest_database_account,
+        database_accounts_list=database_account_list,
+        AZURE_SUBSCRIPTION_ID=subscription_id,
+        azure_update_tag=azure_update_tag,
+    )
+
+
+@timeit
+def sync_database_account_data_resources(
+        neo4j_session: neo4j.Session, subscription_id: str, database_account_list: List[Dict], azure_update_tag: int,
+) -> None:
+    """
+    Loading all the resources (like cors policy, failover policy, private endpoint connections, virtual
+    network rules and locations) that we get as a part of the database account response itself.
+    """
+    for database_account in database_account_list:
+        if 'cors' in database_account and len(database_account['cors']) > 0:
+            db_account = transform_cosmosdb_cors_policy(database_account)
+            _load_cosmosdb_cors_policy(neo4j_session, db_account, azure_update_tag)
+        if 'failover_policies' in database_account and len(database_account['failover_policies']) > 0:
+            _load_cosmosdb_failover_policies(neo4j_session, database_account, azure_update_tag)
+        if 'private_endpoint_connections' in database_account and len(
+                database_account['private_endpoint_connections'],
+        ) > 0:
+            _load_cosmosdb_private_endpoint_connections(neo4j_session, database_account, azure_update_tag)
+        if 'virtual_network_rules' in database_account and len(database_account['virtual_network_rules']) > 0:
+            _load_cosmosdb_virtual_network_rules(neo4j_session, database_account, azure_update_tag)
+
+        locations = []
+        # Extracting every location
+        if 'write_locations' in database_account and len(database_account['write_locations']) > 0:
+            for loc in database_account['write_locations']:
+                locations.append(loc)
+        if 'read_locations' in database_account and len(database_account['read_locations']) > 0:
+            for loc in database_account['read_locations']:
+                locations.append(loc)
+        if 'locations' in database_account and len(database_account['locations']) > 0:
+            for loc in database_account['locations']:
+                locations.append(loc)
+        # Selecting only the unique location entries
+        loc = [i for n, i in enumerate(locations) if i not in locations[n + 1:]]
```

Comment:
How does this select only uniques? Could we use a set instead?

Reply:
@achantavy - So `locations` is a list of dictionaries with each item having the following sample structure - 
{
                "id": "DA1-eastus",
                "location_name": "East US",
                "document_endpoint": "https://DA1-eastus.documents.azure.com:443/",
                "provisioning_state": "Succeeded",
                "failover_priority": 0,
}
And the plan is to return such unique dicts from the list of locations. Since `locations` is a list of dicts, set() can't be used since it is unhashable. I tried it and got a TypeError and saw that explanation on [stackoverflow](https://stackoverflow.com/questions/27435798/unhashable-type-dict-type-error).

So I used the method #2 on this link - https://www.geeksforgeeks.org/python-removing-duplicate-dicts-in-list/ and it worked like a peach! Tested this code and it works :)

Reply:
Are all the items unique by their `"id"`? Here's a nicer way you could get the unique items, borrowing from the third example in your link.
```
>>> test_list = [{'id' : i} for i in range(3)] * 2
>>> test_list
[{'id': 0}, {'id': 1}, {'id': 2}, {'id': 0}, {'id': 1}, {'id': 2}]
>>> unique_lists = {x['id'] : x for x in test_list}.values()
>>> unique_lists
dict_values([{'id': 0}, {'id': 1}, {'id': 2}])
>>> test_list = [{'id' : i} for i in range(3)]*2
>>> test_list
[{'id': 0}, {'id': 1}, {'id': 2}, {'id': 0}, {'id': 1}, {'id': 2}]
>>> uniques = {x['id'] : x for x in test_list}.values()
>>> uniques
dict_values([{'id': 0}, {'id': 1}, {'id': 2}])
>>> list(uniques)
[{'id': 0}, {'id': 1}, {'id': 2}]

```

Reply:
The `id` field is not unique. Hence I stuck to the method I used.

Reply:
Can you provide an example?

Reply:
Something like this - You may have this in 'read locations'
{
"id": "DA1-eastus",
"location_name": "East US",
"document_endpoint": "https://DA1-eastus.documents.azure.com:443/",
"provisioning_state": "Succeeded",
**"failover_priority": 0,**
}

And you may get this in 'write locations' -
{
"id": "DA1-eastus",
"location_name": "East US",
"document_endpoint": "https://DA1-eastus.documents.azure.com:443/",
"provisioning_state": "Succeeded",
**"failover_priority": 10,**
}

In this case, they need to be treated as two different "locations" because of the different values in `failover_priority`.

Reply:
I think I understand now. And later on in line 161-170 you do another filtering of this list. I recommend creating separate lists for each of the types "read", "write", and "associated".

For example
```
write_locations = database_account.get('write_locations', [])
```

And then I think for your `_load_..._locations()` you might also use `UNWIND`.

Reply:
Done!